Olá Neyllor

Meu nome é Ramon e te ajudarei neste projeto. Fico feliz em rever seu projeto hoje.

Ao ao longo do texto farei algumas observações sobre melhorias no código e também farei comentários sobre suas percepções sobre o assunto. Estarei aberta a feedbacks e discussões sobre o tema.

Você encontrará meus comentários abaixo - **por favor, não os mova, modifique ou exclua**.

Você pode encontrar meus comentários em caixas verdes, amarelas ou vermelhas como esta:

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Sucesso. Tudo foi feito corretamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Observações. Algumas recomendações.
</div>

<div class="alert alert-block alert-danger">

<b>Comentário do revisor: </b> <a class="tocSkip"></a>
    
Precisa de correções. O bloqueio requer algumas correções. O trabalho não pode ser aceito com os comentários em vermelho.
</div>

Você pode me responder usando isso:

<div class="alert alert-block alert-info">
<b>Resposta do Aluno.</b> <a class="tocSkip"></a>
</div>


# Analise de comportamento de clientes

Com o objetivo de desenvolver um modelo que possa analisar o comportamento do cliente e recomendar um dos planos mais recentes da Megaline: Smart ou Ultra, serão realizado alguns estudos para indentificar qual seria o tipo de modelo com maior acurácia após ser aplicado no conjunto de dados fornecido. 

## Iniciação 

### Importando bibliotecas

In [1]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

### Carregando dataset

In [2]:
df = pd.read_csv('/datasets/users_behavior_upd.csv')
df.head()

calls  minutes  messages   mb_used  is_ultimate
0   40.0   311.90      83.0  19915.42            0
1   85.0   516.75      56.0  22696.96            0
2   77.0   467.66      86.0  21060.45            0
3  106.0   745.53      81.0   8437.39            1
4   66.0   418.74       1.0  14502.75            0

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   calls        3214 non-null   float64
 1   minutes      3214 non-null   float64
 2   messages     3214 non-null   float64
 3   mb_used      3214 non-null   float64
 4   is_ultimate  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

Ao observar os dados chegou-se a conclusão que os mesmos não apresentam falhas que possam prejudicar na criação dos modelos.

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

## Escolhendo o modelo

### Dividindo os dados em conjunto de treinamento, conjunto de validação e conjunto de teste

In [4]:
#Dividindo os dados em 60% para treinamento e 40% pra validação e teste
df_train, df_valid = train_test_split(df, test_size=0.4, random_state=12345)
#Dividindo os dados de validação e teste em 20%, do total, para cada um deles
df_valid, df_test = train_test_split(df_valid, test_size=0.5, random_state=12345)

features_train = df_train.drop(['is_ultimate'], axis = 1)
target_train = df_train['is_ultimate']

features_valid = df_valid.drop(['is_ultimate'], axis = 1)
target_valid = df_valid['is_ultimate']

features_test = df_test.drop(['is_ultimate'], axis = 1)
target_test = df_test['is_ultimate']

print(features_train.shape)
print(target_train.shape)
print(features_valid.shape)
print(target_valid.shape)
print(features_test.shape)
print(target_test.shape)

(1928, 4)
(1928,)
(643, 4)
(643,)
(643, 4)
(643,)


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

### Experimentando modelos

#### Decion Tree

In [5]:
for depth in range(1, 6):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model.fit(features_train, target_train)
    predictions_valid = model.predict(features_valid)
    print('max_depth =', depth, ': ', end='')
    print(accuracy_score(target_valid, predictions_valid))

max_depth = 1 : 0.7542768273716952
max_depth = 2 : 0.7822706065318819
max_depth = 3 : 0.7853810264385692
max_depth = 4 : 0.7791601866251944
max_depth = 5 : 0.7791601866251944


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

#### Floresta Aleatória

In [6]:
best_score = 0
best_est = 0
for est in range(1, 11):
    model = RandomForestClassifier(random_state=12345, n_estimators=est)
    model.fit(features_train, target_train)
    score = model.score(features_valid, target_valid)
    if score > best_score:
        best_score = score
        best_est = est

print("A acurácia do melhor modelo no conjunto de validação (n_estimators = {}): {}".format(best_est, best_score))

final_model = RandomForestClassifier(random_state=12345, n_estimators=10)
final_model.fit(features_train, target_train)

A acurácia do melhor modelo no conjunto de validação (n_estimators = 10): 0.7853810264385692


RandomForestClassifier(n_estimators=10, random_state=12345)

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

#### Regressão Logística

In [7]:
model = LogisticRegression(random_state=54321, solver='liblinear')
model.fit(features_train, target_train)
score_train = model.score(features_train, target_train)
score_valid = model.score(features_valid, target_valid)

print("Acurácia do modelo de regressão logística no conjunto de treinamento:", score_train)
print("Acurácia do modelo de regressão logística no conjunto de validação:", score_valid)

Acurácia do modelo de regressão logística no conjunto de treinamento: 0.7510373443983402
Acurácia do modelo de regressão logística no conjunto de validação: 0.7573872472783826


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

Após analisar os três modelos, observa-se que os modelos Decision Tree com depth 3 e Floresta Aleatória com 10 árvores possuem a mesma acurácia, 0.7853810264385692. 

O modelo de Regressão Logística foi o que apresentou menor acurácia, 0.7573872472783826.

Por este montivo decidiu-se utilizar o modelo Floresta Aleatória no conjunto de teste.

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

### Aplicando o modelo ao conjunto de teste

In [8]:
# Aplicando modelo
model = RandomForestClassifier(random_state=12345, n_estimators=10)
model.fit(features_test, target_test)
score_test = model.score(features_test, target_test)

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

In [9]:
# Criando tabela final
df_model_final = features_test.merge(right= target_test, on= features_test.index, how= 'outer')
df_model_final.head(10)


key_0  calls  minutes  messages   mb_used  is_ultimate
0    160   61.0   495.11       8.0  10891.23            0
1   2498   80.0   555.04      28.0  28083.58            0
2   1748   87.0   697.23       0.0   8335.70            0
3   1816   41.0   275.80       9.0  10032.39            0
4   1077   60.0   428.49      20.0  29389.52            1
5    791   11.0    58.54       5.0   7085.97            1
6   1024  100.0   709.40      17.0  16964.33            0
7    404   30.0   169.02      13.0  16626.53            0
8   1723   63.0   377.43       0.0  16663.48            0
9    153   72.0   457.21       0.0  13976.58            0

In [10]:
print("Acurácia do modelo Floresta Aleatória no conjunto de teste:", score_test)

Acurácia do modelo Floresta Aleatória no conjunto de teste: 0.9782270606531882


<s><div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Você precisa reportar o resultado da métrica score no teste conforme fez anteriormente
</div>

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

### Conclusões

Após realizar os testes com os modelos, chegou-se a coclusão que o modelo de Floresta Aleatória seria o melhor a ser aplicado neste conjunto de dados uma vez que o mesmo possuia a maior acurácia. Ao final da execução do modelo obtivemos uma Acurácia de 0.9782270606531882.

<s>
<div class="alert alert-block alert-success">
<b>Comentário Geral do Revisor</b> <a class="tocSkip"></a>

Obrigado por enviar seu projeto. 
    
Parabéns, você fez um bom trabalho. 
    
<span class="text-danger">Só há um ponto para correção: calcular a métrica score no teste.</span>
 

Aguardo nova submissão    
Bom trabalho.
</div>



<div class="alert alert-block alert-success">
<b> Comentário Geral do revisor: </b> <a class="tocSkip"></a>

Parabéns. Você finalizou mais uma etapa.
</div>